Feat/2658 add argilla training module for openai with several bug fixes

davidberenstein1957 opened the pull request:
# Description

Updated the argilla.training integration

Closes #2658
Closes #2665 
Closes #2659 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)
- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

argilla/tests/training/*

**Checklist**

- [ ] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

davidberenstein1957 committed:
chore: added not implemented methods to datasetclasses for each task for opanai support

davidberenstein1957 committed:
chore: added support for training prep with openai for textcat #2658

davidberenstein1957 committed:
chore: added context w.r.t. sep and whitespace

davidberenstein1957 committed:
chore: added support for text2text openai

davidberenstein1957 committed:
fix: remove version openai requirements

davidberenstein1957 committed:
fix: added whitespace for text2text completion

davidberenstein1957 committed:
chore: added token classification support openai

davidberenstein1957 committed:
chore: added tests for the openAI integration

davidberenstein1957 committed:
chore: added initial outline for the argilla.training OpenAI module

davidberenstein1957 committed:
chore: added ArgillaTrainingSkeleton Superclass to show general structure and hold init args

davidberenstein1957 committed:
fix: minor fixes w.r.t. initialization

davidberenstein1957 committed:
fix: update default epoch recommendation

davidberenstein1957 committed:
chore: added `settings` param to `prepare_for_training` #2689

davidberenstein1957 committed:
fix: improved completion formatting for openai

davidberenstein1957 committed:
fix: resolved testing with new config and inheritance

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -400,6 +403,24 @@ def prepare_for_training(
 
         """
+        if self._RECORD_TYPE == TextClassificationRecord:
+            print("TextClassificationRecord")
```
```suggestion
```

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -450,16 +471,25 @@ def prepare_for_training(
+                    return self._prepare_for_training_with_openai(records=shuffled_records)
+                else:
+                    raise NotImplementedError(
+                        f"Framework {framework} is not supported. Choose from:" f" {list(Framework)}"
```
```suggestion
                        f"Framework {framework} is not supported. Choose from: {list(Framework)}"
```

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -588,7 +638,7 @@ def from_datasets(
             >>> import datasets
             >>> ds = datasets.Dataset.from_dict({
-            ...     "inputs": ["example"],
+            ...     "inputs": ["exxample"],
```
Is this right?

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -823,29 +875,93 @@ def _prepare_for_training_with_spark_nlp(self, records: List[Record]) -> "pandas
+        self.__all_labels__()  # verify that all labels are strings
+
+        if len(self._records) <= len(self._SETTINGS.label_schema) * 100:
+            _LOGGER.warning("OpenAI recommends at least 100 examples for training a classification model.")
```
```suggestion
            _LOGGER.warning("OpenAI recommends at least 100 examples per class for training a classification model.")
```
I assume you mean class-wise? That matches the code.

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -823,29 +875,93 @@ def _prepare_for_training_with_spark_nlp(self, records: List[Record]) -> "pandas
+        all_labels.sort()
+
+        _LOGGER.warning(
+            f"""No label schema provided. Using all_labels: TextClassificationSettings({all_labels}). We recommend providing a `TextClassificationSettings()` or setting it `rg.configure_dataset_settings()`/`rg.load_dataset_settings()` to ensure reproducibility."""
```
```suggestion
            f"""No label schema provided. Using all_labels: TextClassificationSettings({all_labels}). We recommend providing a `TextClassificationSettings()` or setting `rg.configure_dataset_settings()`/`rg.load_dataset_settings()` to ensure reproducibility."""
```
I think the line is also a bit too long

davidberenstein1957 committed:
chore: updated `CHANGELOG.md`

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -823,29 +875,93 @@ def _prepare_for_training_with_spark_nlp(self, records: List[Record]) -> "pandas
+        )
+        return TextClassificationSettings(all_labels)
+
+    def __all_labels__(self):
```
```suggestion
    def __all_labels__(self):
```
Normally, this "dunder" naming scheme is only used for builtin Python stuff. `_all_labels` is more commonly used for a "private" method.

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -1057,13 +1173,65 @@ def _prepare_for_training_with_spark_nlp(self, records: List[Record]) -> "pandas
+        all_labels = list(all_labels)
+        all_labels.sort()
+        _LOGGER.warning(
+            f"""No label schema provided. Using all_labels: TokenClassificationSettings({all_labels}). We recommend providing a `TokenClassificationSettings()` or setting it `rg.configure_dataset_settings()`/`rg.load_dataset_settings()` to ensure reproducibility."""
```
```suggestion
            f"""No label schema provided. Using all_labels: TokenClassificationSettings({all_labels}). We recommend providing a `TokenClassificationSettings()` or setting `rg.configure_dataset_settings()`/`rg.load_dataset_settings()` to ensure reproducibility."""
```
idem

tomaarsen commented on src/argilla/training/base.py:
```diff
@@ -205,3 +228,39 @@ def save(self, output_dir: str):
         self._trainer.save(output_dir)
+
+
+class ArgillaTrainerSkeleton(object):
```
```suggestion
class ArgillaTrainerSkeleton(object):
```
If you want, you can turn this into an ABC subclass. Then you can use `@abstractmethod` before the yet-to-be-implemented methods, and you don't need the raise NotImplementedError, as any class that has an abstract method that wasn't overridden can't be initialized.
The abc stuff is just from the Python standard library

tomaarsen commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,221 @@
+        **kwargs,
+    ):
+        """
+        Updates the `setfit_model_kwargs` and `setfit_trainer_kwargs` dictionaries with the keyword
```
Mentions of setfit here

tomaarsen commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,221 @@
+
+    def train(self, output_dir: str = None):
+        """
+        We create a SetFitModel object from a pretrained model, then create a SetFitTrainer object with
```
Mentions of setfit

tomaarsen commented on src/argilla/training/spacy.py:
```diff
@@ -15,18 +15,19 @@
 
 if TYPE_CHECKING:
-    import spacy
+    pass
```
You can just remove `if TYPE_CHECKING` if it's not used

tomaarsen commented on src/argilla/training/spacy.py:
```diff
@@ -102,21 +99,24 @@ def __init__(
             try:
                 spacy.prefer_gpu(self.gpu_id)
-            except:
+            except Exception:
```
Maybe we can make this more specific?

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -197,11 +190,13 @@ def preprocess_datasets(self):
         def text_classification_preprocess_function(examples):
-            tokens = self._tokenizer(examples["text"], truncation=True, padding="max_length", max_length=512)
+            tokens = self._transformers_tokenizer(
+                examples["text"], truncation=True, padding="max_length", max_length=512
```
Do we have to hardcode 512?

tomaarsen reviewed (commented):
Lots of small improvements, I like it. I found a few nits that should be pretty easy to resolve.

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -81,37 +80,31 @@ def __init__(self, dataset, record_class, multi_label: bool = False, model: str
-        else:
-            raise NotImplementedError("rg.Text2TextRecords are not supported yet.")
+                    self._eval_dataset = _apply_column_mapping(self._eval_dataset, columns_mapping)
+        print(self._train_dataset)
```
Leftover print

tomaarsen commented on src/argilla/client/datasets.py:
```diff
@@ -450,16 +471,25 @@ def prepare_for_training(
+                        f"Framework {framework} is not supported. Choose from:" f" {list(Framework)}"
+                    )
         else:
             raise NotImplementedError(f"Framework {framework} is not supported. Choose from:" f" {list(Framework)}")
```
This is formatted like:
```python
NotImplementedError: Framework span_marker is not supported. Choose from: [<Framework.TRANSFORMERS: 'transformers'>, <Framework.SETFIT: 'setfit'>, <Framework.SPACY: 'spacy'>, <Framework.SPAN_MARKER: 
'span_marker'>, <Framework.SPARK_NLP: 'spark-nlp'>, <Framework.OPENAI: 'openai'>]
```
Not ideal

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -47,32 +58,20 @@ def __init__(self, dataset, record_class, multi_label: bool = False, model: str
         else:
             self.device = "cpu"
```
```suggestion
```
There's already a `self.device = "cpu"` before

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -302,7 +299,6 @@ def compute_metrics(p):
         return func
 
-    # @require_version("setfit", "0.6")
     def train(self, output_dir: str = None):
```
```suggestion
    def train(self, output_dir: str):
```
If it's required (see the ValueError below), then can't we just require it in the signature like this?

tomaarsen commented on src/argilla/training/transformers.py:
I can't make a comment down there, but on this section:
https://github.com/argilla-io/argilla/blob/2a1aac5c6f5074b4551e52ffdaa9aabffec7963d/src/argilla/training/transformers.py#L162-L173

* setfit is mentioned in the docstring
* the formatting is off - the signature can just fit on one line.
* The two update lines can be:
  ```python
  self.trainer_kwargs.update(filter_allowed_args(TrainingArguments.__init__, **self.kwargs))
  ```
  with effectively the same behaviour, I believe.

davidberenstein1957 committed:
Apply suggestions from code review

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
fix: resolved initial comments Tom

davidberenstein1957 committed:
fix: fixed tests for transformers training

davidberenstein1957 committed:
fix: updated settings config OpenAI

davidberenstein1957 committed:
chore: updated settings configuration

tomaarsen committed:
Merge branch 'develop' of https://github.com/argilla-io/argilla into feat/2658-add-argilla-training-module-for-openai

tomaarsen committed:
Resolve import error in tests

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,223 @@
+            self._logger.warning("Seed is not supported for OpenAI. Ignoring seed for training.")
+
+        if self._model is None:
+            self._model = "curie"
```
Maybe we can default `ada`? Just pointing it out because it's the fastest and cheaper OpenAI model to fine-tune. See https://platform.openai.com/docs/models/gpt-3

davidberenstein1957 replied:
Hi, "curie" is the default offered by the openai API, so I wanted to align with that.

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,223 @@
+        self.model_kwargs["validation_file"] = validation_file
+        self.model_kwargs["model"] = model
+        if isinstance(self._settings, TextClassificationSettings):
+            self.model_kwargs["n_epochs"] = 4
```
```suggestion
            self.model_kwargs["n_epochs"] = n_epochs or 4
```

tomaarsen replied:
Good find! Thanks for this one (and the other similar one)

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,223 @@
+        if isinstance(self._settings, TextClassificationSettings):
+            self.model_kwargs["n_epochs"] = 4
+        else:
+            self.model_kwargs["n_epochs"] = 2
```
```suggestion
            self.model_kwargs["n_epochs"] = n_epochs or 2
```

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,223 @@
+            kwargs = {"logprobs": len(self._settings.label_schema), "max_tokens": 1}
+
+        for entry in text:
+            prompt = entry + self._separator
```
Here maybe we can just `strip()` the str before adding the separator in case there's any line break in the string.

```suggestion
            prompt = f"{entry.strip()}{self._separator}"
```

tomaarsen replied:
Strip does not remove line breaks unless they are at the start/end of the string. I also think the f-string formatting is a bit overkill if we're only concatenating strings. @davidberenstein1957

alvarobartt replied:
Yes, but since the separator is `\n\n###\n\n` we should be removing at least the line breaks at the end, since the tokenizer counts the line breaks as tokens, and I feel like removing the extra line breaks is more efficient here. Both at the start and at the end, since the rest of the line breaks may be on purpose.

alvarobartt replied:
Also, regarding the f-string I'd say that the performance compared to string concatenation is pretty similar indeed. But we can rely on string concatenation for now!

tomaarsen committed:
Resolve ignored n_epochs in configuration

Co-authored-by: Alvaro Bartolome <alvarobartt@yahoo.com>

alvarobartt committed:
Fix GPU usage & other minor improvements in `ArgillaSpaCyTrainer` (#2667)

# Description

After extensively testing `ArgillaTrainer` as of the latest release for
`spaCy`, I've found some things that had to be
fixed/improved/refactored:

* Since `gpu_id` was not provided in `ArgillaTrainer`, then it was not
provided to `ArgillaSpaCyTrainer` so the GPU could only be used when
using `ArgillaSpaCyTrainer` directly
* `ArgillaSpaCyTrainer.train` was still using `path` as the param
instead of `output_dir`
* Remove unnecessary `requires_version` function calls in
`ArgillaSpaCyTrainer`, and add `requires_version("spacy-transformers")`
just when required
* Improve `gpu_id` param docstrings in both `ArgillaTrainer` and
`ArgillaSpaCyTrainer`

**Type of change**

- [X] Bug fix (non-breaking change which fixes an issue)
- [X] Refactor (change restructuring the codebase without changing
functionality)
- [X] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

- [X] Add `test_init_with_gpu_id` integration tests to check that
providing `gpu_id` to `ArgillaTrainer` with `ArgillaSpaCyTrainer` works
as expected
- [X] Train a token-classification/text-classification model using
`ArgillaSpaCyTrainer` with and without GPU in Google Colab via the
`gpu_id` param being 0 and 1, respectively. _I can upload those either
via sharing the link or just adding those to the tutorials directory if
needed._

**Checklist**

- [X] I have merged the original branch into my forked branch
- [X] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

davidberenstein1957 committed:
fix: updated certificate versioning

davidberenstein1957 committed:
chore: added minimal code cleaning

davidberenstein1957 committed:
fix: removed specifc veriosningh certifi packages docs.yml deps

tomaarsen commented on src/argilla/training/transformers.py:
```diff
@@ -81,37 +78,32 @@ def __init__(self, dataset, record_class, multi_label: bool = False, model: str
+
+            self._model_class = AutoModelForTokenClassification
         else:
             raise NotImplementedError("rg.Text2TextRecord are not supported")
```
```suggestion
            raise NotImplementedError("rg.Text2TextRecord is not supported.")
```

tomaarsen committed:
Feat: Add SpanMarker Argilla Trainer for NER (#2693)

Hello!

## Pull Request overview
* Add [SpanMarker](https://github.com/tomaarsen/SpanMarkerNER) Argilla
Trainer for Named Entity Recognition.

## Details
The SpanMarker Argilla trainer is based on the Transformers Trainer, as
SpanMarker is tightly implemented on top of transformers. However, we
don't need to do the tokenization, data collation or evaluation on the
Argilla side, unlike with Transformers. This makes the SpanMarker
Argilla trainer relatively small.

## Usage
First, we need an annotated dataset:
```python
import argilla as rg
from datasets import load_dataset

dataset = "conll2003"
dataset_ds = load_dataset(
    "conll2003",
    split="train[:1000]",
)
dataset_ds = dataset_ds.rename_column("ner_tags", "tags")
dataset_rb = rg.read_datasets(dataset_ds, task="TokenClassification")

rg.delete(dataset)
rg.log(name=dataset, records=dataset_rb)
```
And then we can use the new Trainer to train with this dataset:
```python
import argilla as rg
from argilla.training.base import ArgillaTrainer

dataset = "conll2003"

trainer = ArgillaTrainer(name=dataset, framework="span_marker", train_size=0.8)
trainer.update_config(
    num_train_epochs=10,
    bf16=True,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=16,
    marker_max_length=128,
    entity_max_length=8,
)
trainer.train(output_dir="tmp_span_marker_train")
```
(You can use lower batch sizes or `model_max_length=256` if you have
memory issues. You can also use `fp16` instead of `bf16` if you get an
error.)

This produces the following logs:

<details><summary>Click to see the logs</summary>

```
[04/13/23 16:25:25] WARNING  WARNING:argilla.client.datasets:No label schema provided. Using all_labels: TokenClassificationSettings(['LOC', 'MISC', 'ORG', 'PER']). We recommend      datasets.py:1222
                             providing a `TokenClassificationSettings()` or setting it `rg.configure_dataset_settings()`/`rg.load_dataset_settings()` to ensure reproducibility.       
[04/13/23 16:25:30] WARNING  WARNING:ArgillaTrainer:ArgillaBaseTrainer info:                                                                                                                base.py:175
                             _________________________________________________________________
                             These baseline params are fixed:
                                 dataset: conll2003
                                 task: DatasetForTokenClassification
                                 multi_label: False
                                 train_size: 0.8
                                 seed: None

                             <class 'argilla.training.span_marker.ArgillaSpanMarkerTrainer'> info:
                             _________________________________________________________________
                             The parameters are configurable via `trainer.update_config()`:
                                 'SpanMarkerModel'
                             pretrained_model_name_or_path: bert-base-cased
                             labels: ['O', 'B-LOC', 'I-LOC', 'B-MISC', 'I-MISC', 'B-ORG', 'I-ORG', 'B-PER', 'I-PER']
                             'Trainer'
                             overwrite_output_dir: False
                             do_train: False
                             do_eval: False
                             do_predict: False
                             evaluation_strategy: epoch
                             prediction_loss_only: False
                             per_device_train_batch_size: 8
                             per_device_eval_batch_size: 8
                             per_gpu_train_batch_size: None
                             per_gpu_eval_batch_size: None
                             gradient_accumulation_steps: 1
                             eval_accumulation_steps: None
                             eval_delay: 0
                             learning_rate: 5e-05
                             weight_decay: 0.01
                             adam_beta1: 0.9
                             adam_beta2: 0.999
                             adam_epsilon: 1e-08
                             max_grad_norm: 1.0
                             num_train_epochs: 3.0
                             max_steps: -1
                             lr_scheduler_type: linear
                             warmup_ratio: 0.0
                             warmup_steps: 0
                             log_level: passive
                             log_level_replica: warning
                             log_on_each_node: True
                             logging_dir: None
                             logging_strategy: steps
                             logging_first_step: False
                             logging_steps: 30
                             logging_nan_inf_filter: True
                             save_strategy: steps
                             save_steps: 500
                             save_total_limit: None
                             save_on_each_node: False
                             no_cuda: False
                             use_mps_device: False
                             seed: 42
                             data_seed: None
                             jit_mode_eval: False
                             use_ipex: False
                             bf16: False
                             fp16: False
                             fp16_opt_level: O1
                             half_precision_backend: auto
                             bf16_full_eval: False
                             fp16_full_eval: False
                             tf32: None
                             local_rank: -1
                             xpu_backend: None
                             tpu_num_cores: None
                             tpu_metrics_debug: False
                             debug:
                             dataloader_drop_last: False
                             eval_steps: None
                             dataloader_num_workers: 0
                             past_index: -1
                             run_name: None
                             disable_tqdm: None
                             remove_unused_columns: True
                             label_names: None
                             load_best_model_at_end: False
                             metric_for_best_model: None
                             greater_is_better: None
                             ignore_data_skip: False
                             sharded_ddp:
                             fsdp:
                             fsdp_min_num_params: 0
                             fsdp_config: None
                             fsdp_transformer_layer_cls_to_wrap: None
                             deepspeed: None
                             label_smoothing_factor: 0.0
                             optim: adamw_hf
                             optim_args: None
                             adafactor: False
                             group_by_length: False
                             length_column_name: length
                             report_to: None
                             ddp_find_unused_parameters: None
                             ddp_bucket_cap_mb: None
                             dataloader_pin_memory: True
                             skip_memory_metrics: True
                             use_legacy_prediction_loop: False
                             push_to_hub: False
                             resume_from_checkpoint: None
                             hub_model_id: None
                             hub_strategy: every_save
                             hub_token: None
                             hub_private_repo: False
                             gradient_checkpointing: False
                             include_inputs_for_metrics: False
                             fp16_backend: auto
                             push_to_hub_model_id: None
                             push_to_hub_organization: None
                             push_to_hub_token: None
                             mp_parameters:
                             auto_find_batch_size: False
                             full_determinism: False
                             torchdynamo: None
                             ray_scope: last
                             ddp_timeout: 1800
                             torch_compile: False
                             torch_compile_backend: None
                             torch_compile_mode: None
                             output_dir: None

                             Using the trainer:
                             _________________________________________________________________
                             `trainer.train(output_dir)` to train to start training. `output_dir` is the directory to save the model automatically.
                             `trainer.predict(text, as_argilla_records=True)` to make predictions.
                             `trainer.save(output_dir)` to save the model manually.
Some weights of the model checkpoint at bert-base-cased were not used when initializing BertModel: ['cls.predictions.transform.dense.bias', 'cls.predictions.bias', 'cls.seq_relationship.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.decoder.weight', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.LayerNorm.weight']
- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
C:\Users\tom\.conda\envs\argilla\lib\site-packages\transformers\optimization.py:395: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  FutureWarning,
{'loss': 0.2601, 'learning_rate': 4.754098360655738e-05, 'epoch': 0.49}
{'loss': 0.066, 'learning_rate': 4.508196721311476e-05, 'epoch': 0.98}                                                                                                                                 
{'eval_loss': 0.03822080418467522, 'eval_overall_precision': 0.8810289389067524, 'eval_overall_recall': 0.6372093023255814, 'eval_overall_f1': 0.7395411605937922, 'eval_overall_accuracy': 0.9204204204204204, 'eval_runtime': 4.5471, 'eval_samples_per_second': 57.619, 'eval_steps_per_second': 3.739, 'epoch': 1.0}                                                                                      
{'loss': 0.0378, 'learning_rate': 4.262295081967213e-05, 'epoch': 1.48}                                                                                                                                
{'loss': 0.029, 'learning_rate': 4.016393442622951e-05, 'epoch': 1.97}
{'eval_loss': 0.021275097504258156, 'eval_overall_precision': 0.8808933002481389, 'eval_overall_recall': 0.8255813953488372, 'eval_overall_f1': 0.8523409363745499, 'eval_overall_accuracy': 0.9602102102102102, 'eval_runtime': 4.5328, 'eval_samples_per_second': 57.8, 'eval_steps_per_second': 3.75, 'epoch': 2.0}                                                                                        
{'loss': 0.0169, 'learning_rate': 3.7704918032786885e-05, 'epoch': 2.46}
{'loss': 0.015, 'learning_rate': 3.524590163934427e-05, 'epoch': 2.95}
{'eval_loss': 0.013853945769369602, 'eval_overall_precision': 0.9527363184079602, 'eval_overall_recall': 0.8906976744186047, 'eval_overall_f1': 0.920673076923077, 'eval_overall_accuracy': 0.9774774774774775, 'eval_runtime': 4.4899, 'eval_samples_per_second': 58.353, 'eval_steps_per_second': 3.786, 'epoch': 3.0}                                                                                      
{'loss': 0.01, 'learning_rate': 3.2786885245901635e-05, 'epoch': 3.44}
{'loss': 0.0089, 'learning_rate': 3.0327868852459017e-05, 'epoch': 3.93}
{'eval_loss': 0.013151775114238262, 'eval_overall_precision': 0.948780487804878, 'eval_overall_recall': 0.9046511627906977, 'eval_overall_f1': 0.9261904761904762, 'eval_overall_accuracy': 0.9786036036036037, 'eval_runtime': 4.5472, 'eval_samples_per_second': 57.618, 'eval_steps_per_second': 3.739, 'epoch': 4.0}                                                                                      
{'loss': 0.0056, 'learning_rate': 2.7868852459016392e-05, 'epoch': 4.43}                                                                                                                               
{'loss': 0.0059, 'learning_rate': 2.540983606557377e-05, 'epoch': 4.92}
{'eval_loss': 0.012591547332704067, 'eval_overall_precision': 0.9587378640776699, 'eval_overall_recall': 0.9186046511627907, 'eval_overall_f1': 0.9382422802850356, 'eval_overall_accuracy': 0.9812312312312312, 'eval_runtime': 4.5184, 'eval_samples_per_second': 57.986, 'eval_steps_per_second': 3.762, 'epoch': 5.0}                                                                                     
{'loss': 0.0036, 'learning_rate': 2.295081967213115e-05, 'epoch': 5.41}
{'loss': 0.0044, 'learning_rate': 2.0491803278688525e-05, 'epoch': 5.9}
{'eval_loss': 0.012911035679280758, 'eval_overall_precision': 0.9539951573849879, 'eval_overall_recall': 0.9162790697674419, 'eval_overall_f1': 0.9347568208778173, 'eval_overall_accuracy': 0.9808558558558559, 'eval_runtime': 4.3026, 'eval_samples_per_second': 60.893, 'eval_steps_per_second': 3.951, 'epoch': 6.0}                                                                                     
{'loss': 0.0031, 'learning_rate': 1.8032786885245903e-05, 'epoch': 6.39}
{'loss': 0.0024, 'learning_rate': 1.557377049180328e-05, 'epoch': 6.89}
{'eval_loss': 0.010593990795314312, 'eval_overall_precision': 0.9567307692307693, 'eval_overall_recall': 0.9255813953488372, 'eval_overall_f1': 0.9408983451536642, 'eval_overall_accuracy': 0.9838588588588588, 'eval_runtime': 4.5172, 'eval_samples_per_second': 58.001, 'eval_steps_per_second': 3.763, 'epoch': 7.0}                                                                                     
{'loss': 0.0018, 'learning_rate': 1.3114754098360657e-05, 'epoch': 7.38}
{'loss': 0.0025, 'learning_rate': 1.0655737704918032e-05, 'epoch': 7.87}
{'eval_loss': 0.010297469794750214, 'eval_overall_precision': 0.9478672985781991, 'eval_overall_recall': 0.9302325581395349, 'eval_overall_f1': 0.9389671361502349, 'eval_overall_accuracy': 0.9846096096096096, 'eval_runtime': 4.4597, 'eval_samples_per_second': 58.748, 'eval_steps_per_second': 3.812, 'epoch': 8.0}                                                                                     
{'loss': 0.0011, 'learning_rate': 8.196721311475409e-06, 'epoch': 8.36}                                                                                                                                
{'loss': 0.0009, 'learning_rate': 5.737704918032787e-06, 'epoch': 8.85}
{'eval_loss': 0.009832620620727539, 'eval_overall_precision': 0.9478672985781991, 'eval_overall_recall': 0.9302325581395349, 'eval_overall_f1': 0.9389671361502349, 'eval_overall_accuracy': 0.9846096096096096, 'eval_runtime': 4.6756, 'eval_samples_per_second': 56.036, 'eval_steps_per_second': 3.636, 'epoch': 9.0}                                                                                     
{'loss': 0.0015, 'learning_rate': 3.278688524590164e-06, 'epoch': 9.34}                                                                                                                                
{'loss': 0.0013, 'learning_rate': 8.19672131147541e-07, 'epoch': 9.84}
{'eval_loss': 0.010347267612814903, 'eval_overall_precision': 0.9457547169811321, 'eval_overall_recall': 0.9325581395348838, 'eval_overall_f1': 0.9391100702576113, 'eval_overall_accuracy': 0.9846096096096096, 'eval_runtime': 4.7101, 'eval_samples_per_second': 55.625, 'eval_steps_per_second': 3.609, 'epoch': 10.0}                                                                                    
{'train_runtime': 297.0633, 'train_samples_per_second': 32.855, 'train_steps_per_second': 2.053, 'train_loss': 0.023517039891515597, 'epoch': 10.0}                                                   
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 610/610 [04:57<00:00,  2.05it/s] 
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 17/17 [00:04<00:00,  3.84it/s] 
[04/13/23 16:30:34] INFO     INFO:ArgillaSpanMarkerTrainer:{'eval_loss': 0.010347267612814903, 'eval_overall_precision': 0.9457547169811321, 'eval_overall_recall':                  span_marker.py:133
                             0.9325581395348838, 'eval_overall_f1': 0.9391100702576113, 'eval_overall_accuracy': 0.9846096096096096, 'eval_runtime': 4.6647,
                             'eval_samples_per_second': 56.166, 'eval_steps_per_second': 3.644, 'epoch': 10.0}
```
</details>

In short, I trained to 0.939 eval F1 on CoNLL03 in 5 minutes.

### Type of change

- [x] New feature (non-breaking change which adds functionality)

### How Has This Been Tested
Tests still need to be written. I'll be working on this - but I'll
publish this as a draft already so it's available for reviews already.

<!--
**Checklist**

- [ ] I have merged the original branch into my forked branch
- [ ] I added relevant documentation
- [ ] follows the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my
feature works
- [ ] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)
-->

- Tom Aarsen

---------

Co-authored-by: David Berenstein <david.m.berenstein@gmail.com>

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -0,0 +1,225 @@
+
+        if self._train_dataset is not None and self.model_kwargs["training_file"] is None:
+            self.model_kwargs["training_file"] = self.upload_dataset_to_openai(self._train_dataset, "data_train.jsonl")
+        if self._eval_dataset is not None and self.model_kwargs["validation_file"] is None:
```
I think that the `validation_file` is mandatory when `compute_classification_metrics=True`, @davidberenstein1957 can you confirm?

If so, then we should handle this to show a warning and set the `compute_classification_metrics` to `False` in case the `validation_file` is not provided

alvarobartt commented on src/argilla/datasets/__init__.py:
```diff
@@ -38,3 +63,19 @@ def configure_dataset(name: str, settings: Settings, workspace: Optional[str] =
+        settings: The dataset settings
+        workspace: The workspace name where the dataset will belongs to
+    """
+    _LOGGER.warning("This method is deprecated. Use configure_dataset_settings instead.")
```
Does it make sense to set `logging.captureWarnings(capture=True)` to log this warning as a `DeprecationWarning` using `warnings`? More information at https://docs.python.org/3/library/logging.html#logging.captureWarnings

tomaarsen replied:
This is a wider issue, see also #2456.

alvarobartt replied:
Ok, I'll add this comment there then!

davidberenstein1957 committed:
Apply suggestions from code review

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
chore: #2658 included tests for openai

alvarobartt commented on src/argilla/training/openai.py:
```diff
@@ -164,6 +176,18 @@ def train(self, output_dir: str = None):
         self._logger.info(response)
         self.finetune_id = response.id
 
+        while True:
```
I wouldn't recommend to create a `while True` loop, since the fine-tuning may last too long and the function would block the execution...

So I'd suggest just to return the fine-tune ID and then show a warning to use `openai api fine_tunes.follow -i <YOUR_FINE_TUNE_JOB_ID>` to track the progress, because at least from my experience when I created https://github.com/alvarobartt/opentrain is that the fine-tune takes way too long.

alvarobartt commented on tests/training/test_openai.py:
```diff
@@ -0,0 +1,52 @@
+from argilla.training import ArgillaTrainer
+
+FRAMEWORK = "openai"
+MODEL = "cure"
```
```suggestion
MODEL = "curie"
```

alvarobartt reviewed (commented):
Just some minor comments cc. @davidberenstein1957

tomaarsen reviewed (commented):
`openai` isn't in the CI dependencies yet.

davidberenstein1957 committed:
fix: add openai to dependencies

davidberenstein1957 committed:
chore: add `CLI` support to trainer #2809

davidberenstein1957 committed:
chore: added CLI support `ArgillaTrainer`
chore: switched `click` to `typer`

davidberenstein1957 committed:
chore: updated changelog.md

davidberenstein1957 committed:
chore: updated fine-tuning readme.md files

davidberenstein1957 committed:
chore: updated CLI usage docker .sh files

davidberenstein1957 committed:
chore: updated documentation on CLI usage

davidberenstein1957 committed:
chore: added docs on CLI support trainer

davidberenstein1957 committed:
Merge branch 'develop' into feat/2658-add-argilla-training-module-for-openai

davidberenstein1957 committed:
Update tests/training/test_openai.py

Co-authored-by: Alvaro Bartolome <alvarobartt@yahoo.com>

davidberenstein1957 committed:
fix: resolved tests due to bug in click

davidberenstein1957 committed:
fix: resolved logging to show updated params

davidberenstein1957 committed:
chore: renames `cleanup` to `train_with_cleanup``
chore: removed while-loop openai trainer

tomaarsen commented on pyproject.toml:
```diff
@@ -44,7 +44,9 @@ dependencies = [
-    "rich <= 13.0.1"
+    "rich <= 13.0.1",
+    # for CLI
+    "typer >= 0.6.0, < 0.8.0"
```
```suggestion
    "typer >= 0.6.0, < 1.0.0"
```
Edited to add the suggestion from Alvaro below.

Perhaps we can be more flexible on this upper bound now that 0.9.0 is out?

alvarobartt replied:
According to the `0.9.0` just the default/recommended argument definition changed, but everything's backwards compatible, so we can probably just `>=0.6.0,<1.0.0`?

davidberenstein1957 committed:
Update pyproject.toml

Co-authored-by: Tom Aarsen <37621491+tomaarsen@users.noreply.github.com>

davidberenstein1957 committed:
fix: model_init for spacy
fix: replaced pass for docstrings abc
chore: added tests spacy wo training
chore: added tests transformers wo training
chore: added specific setfit versioning
